api/app/resources/endpoints.py

```diff
@@ -0,0 +1,252 @@
+            # If limit specified by client isn't number type, ignore
+            # that and default to 20
+            limit = 20
+
+        if limit:
+            # if limit is greater than maximum, default to 100
+            if limit > 100:
+                Limit.limit = 100
+            elif limit < 1:
+            # if limit is <= 0, default to 20
+                Limit.limit = 20
+            else:
+                Limit.limit = limit
+
+        current_user = User.verify_auth_token(request.headers.get('token'))
+
+        # the "search bucketlist by name" parameter
+        search = False
+        qs = request.args.get('q')
+        if qs:
+            search = True
+        try:
+            page = int(request.args.get('page', 1))
+        except ValueError:
+            page = 1
+
+        if search:
+            query = Bucketlist.query.filter_by(
+                name=qs)
+            if query.count() == 0:
+                return {"message": "No search results"}
+            else:
+                info = {"message": "Search results"}
+                return [info, marshal(query.first(), bucketlist_fields)]
+
+        bucketlists = Bucketlist.query.all()
+        pagination = Pagination(per_page=Limit.limit, page=page,
+                                total=len(bucketlists), search=search,
+                                record_name='bucketlists')
+        info = {"Pagination Details": {
+            "results": pagination.endpoint,
+            "per_page": pagination.per_page,
+            "page": pagination.page,
+            "total": pagination.total}}
+        return [info, marshal(bucketlists, bucketlist_fields)]
+
+
+    @auth.login_required
+    def post(self):
+        """POST endpoint"""
+        args = self.parser.parse_args()
+        name = args['name']
+        created_by = args['created_by']
+        new_bucket_list = Bucketlist(name=name, created_by=created_by)
+        name = new_bucket_list.name
+        db.session.add(new_bucket_list)
+        db.session.commit()
+        return {'message': "'{}' successfully created".format(name)}, 201
+
+
+class SingleBucketlist(Resource):
+    """
+    A Single Bucketlist [/bucketlists/<>]
+    """
+
+    def __init__(self):
+        """Instantiate route request parameters"""
+        self.parser = reqparse.RequestParser()
+        self.parser.add_argument('name', type=str,
+                                 help='Enter name of Bucketlist',
+                                 location='json')
+        super(SingleBucketlist, self).__init__()
+
+    @auth.login_required
+    @marshal_with(bucketlist_fields)
+    def get(self, id):
+        """GET endpoint"""
+        bucketlist = Bucketlist.query.filter_by(id=id).first()
+        return bucketlist
+
+    @auth.login_required
+    def put(self, id):
+        """PUT endpoint"""
+        args = self.parser.parse_args()
+        name = args['name']
+        bucket_list = Bucketlist.query.filter_by(id=id).first()
```

Comment:
**Code Quality: Critical**

There is a possibility of an exception being thrown if you don't check to make sure the record you're trying to select actually exists.